### PR TITLE
Fetch fore sce

### PR DESCRIPTION
**Pull request by msleckman**

### CODE REVIEW

File: 1_fetch/src/fetch_FORESCE.R
Comment (msleckman) [inline 1_fetch/src/fetch_FORESCE.R:21]: Should years be a function param? Leaning against because requires converting a list of years to regex, and I did not find a way non-regex way to subset the downloaded tif file names.

File: 1_fetch/src/fetch_FORESCE.R
Comment (msleckman) [inline 1_fetch/src/fetch_FORESCE.R:1]: Not that this is rather generic function and possible redundant to what is already in fetch. Adding point to consider combining `sb::item_file_download()` functions at a later point.

File: 1_fetch/src/fetch_FORESCE.R
Comment (msleckman) [inline 1_fetch/src/fetch_FORESCE.R:25]: Do we prefer to return path to folder in which tif files reside? Or rather return path to specified files. Kept the later for now.

**Comment by msleckman** [comment]
### Raster Processing per PRMS Catchment

This branch now ready for review. 

Data reviewed: 

* Historical Land Cover drb (res 30 by 30)
image

* [Bock road Salt](https://www.sciencebase.gov/catalog/item/5b15a50ce4b092d9651e22b9) (res 1000 by 1000) 
image


Added two functions and targets to pipeline so far:

* Targets names:
  * `p1_backcasted_LC` (file)
  * `p2_LC_per_catchment` (df) 
* Functions: 
1. `fetch_FORESCE.R`: includes `download_backcasted_LC()` , pulls the backcasted data from sciencebase and selects just the relevant years (1960 - 2000)

2. `raster_per_PRMS.R`: includes `raster_to_catchment_polygons()`,  to produce summary of raster values per PRMS catchment. Function adapted to take discrete and non-discrete rasters. Used an adapted version of `rbind.fill()` function - named it `rbind.fill.list()`  - that can rbind and fill with a list as input ( `plyr::rbind.fill()` only takes dataframes)

`raster_to_catchment_polygons()` computes these steps: 

1. checks geometries and crs of raster & vector. Note that the p1_catchments_sf has invalid geometries
2. crops raster to the vector geometry (sf object)

> example crop of road salt raster (crop only clipped rectangle). 

image

3. Extract and summarize raster pixels per polygon

* If discrete raster (Land Cover) --> calculate proportion of raster class in polygon. Output is the catchment spatial df with added cols (lcClass0...).
 Subset sample:

```
> p2_LC_per_catchment[,10:17] %>% st_drop_geometry() %>% head()

region Shape_Length Shape_Area ID    lcClass0   lcClass1  lcClass2    lcClass3
1     02    64862.502   61982133  1 0.004327307 0.06630364 0.1440209 0.029303710
2     02     8368.805    1566006  2          NA         NA        NA          NA
3     02    21943.396    4896969  3          NA         NA        NA          NA
4     02   106081.682  158390574  4 0.015052163 0.09025047 0.2445962 0.021774212
5     02    47627.926   22488083  5          NA 0.08537464 0.2690922 0.008525456
6     02    11826.650    2199987  6          NA 0.05389955 0.1784402 0.077991017

```

* If non-discrete (salt) --> calculate sum of raster values. Output is the catchment spatial df with 1 added col (rd_saltX1992)
```
> head(df[,7:14] %>% st_drop_geometry())
hru_lat   hru_area hru_segment region Shape_Length Shape_Area ID rd_saltX1992
1 38.77222 15316.1187        1263     02    64862.502   61982133  1   138155.778
2 38.82182   386.9685        1264     02     8368.805    1566006  2        0.000
3 38.81343  1210.0674        1263     02    21943.396    4896969  3     3469.941
4 38.76383 39139.1631        1264     02   106081.682  158390574  4   212863.080
5 38.83563  5556.9263        1264     02    47627.926   22488083  5    29178.849
6 38.90042   543.6286        1260     02    11826.650    2199987  6     2177.043

```

**Comments/Questions:**

* Raster processing (cropping, and extracting) was done using `terra` instead of `raster` as the processing is much faster. Sample speed testing below (ellapsed = seconds passed): 

```
> raster <- raster(p1_backcasted_LC[1])
> vector_sf <- p1_catchments_sf %>% mutate(id = row_number()) 
> SpatRaster <- rast(p1_backcasted_LC[1])
> SpatVector <- p1_catchments_sf %>% mutate(id = row_number()) %>% vect()
> 
> system.time({raster_per_polygon_list <- raster::extract(raster, vector_sf[1:5,])})

user  system elapsed 
 12.251   1.556  13.856 
Warning message:
In .local(x, y, ...) :
  Transforming SpatialPolygons to the crs of the Raster

> system.time({raster_per_polygon_list <- terra::extract(SpatRaster, SpatVector[1:5,])})
 
user  system elapsed 
0.945   0.035   0.981 
  
```

* For non-discrete raster, the summary function is set to `sum`. Is this what we want per PRMS catchment for non-discrete rasters such as rd salt? or do we want mean? 

* For LC raster, the calculation produces a proportion calculation - # pixels/ total pixels per catchment. Need more calculation steps to get final result of lc class area/total area. 

* So far, this step has only been tested on individual `.tif` files. Need to compute this iteratively in the pipeline to get all relevant decades or years appended. 

**Still to do:**
* Fetch road salt data --> not part of this issue so can create a different branch for adding this 
* add final calc step for lc area 
* Add mapping in pipeline to loop through tif files 
* Quickly compare `terra::crop()` with `terra::mask`
* New issue combining hist LC to NLCD land cover.

File: 2_process/src/raster_per_PRMS.R
Comment (msleckman) [inline 2_process/src/raster_per_PRMS.R:1]: Adapted function to `rbind.fill()` to take in list of tables as inputs 
sample of input x: 

```
>   print(raster_per_polygon[1:5])
[[1]]
x
          0           1           2           3           4           5           6           8           9          10          11          12          13 
0.004327307 0.066303638 0.144020910 0.029303710 0.027546649 0.025804109 0.077775358 0.039294271 0.023916358 0.018645175 0.007768823 0.003470558 0.240499528 
         14          15          16          17          18          19 
0.003165614 0.015029405 0.009540405 0.007507442 0.064677267 0.191403471 

[[2]]
x
           9           10           11           12           13           15           17           18           19 
0.0682339450 0.0028669725 0.0005733945 0.0011467890 0.0108944954 0.0005733945 0.0275229358 0.0303899083 0.8577981651 

[[3]]
x
           9           10           12           13           14           15           16           17           18           19 
0.0874196511 0.0001836547 0.0113865932 0.0029384757 0.0003673095 0.0145087236 0.0056932966 0.0121212121 0.0385674931 0.8268135904 

```

File: _targets.R
Comment (jds485) [inline _targets.R:91]: could include LC in variable name

File: 1_fetch/src/fetch_FORESCE.R
Comment (jds485) [reply 1_fetch/src/fetch_FORESCE.R:27]: path to files seems more consistent with the file format for targets

File: 2_process.R
Comment (jds485) [inline 2_process.R:70]: Is this for only year 2000? `p1_backcasted_LC[1]`

File: 2_process/src/raster_per_PRMS.R
Comment (jds485) [inline 2_process/src/raster_per_PRMS.R:36]: I noticed that some of the PRMS HRU polygons have loops/holes in them (I don't know the technical term for this). Is that the problem that this function fixes?

File: 2_process/src/raster_per_PRMS.R
Comment (jds485) [inline 2_process/src/raster_per_PRMS.R:56]: these processing times can probably be removed for the commit to main

File: 2_process.R
Comment (jds485) [reply 2_process.R:70]: Or is this target missing a `map` over the rasters?

File: 2_process/src/raster_per_PRMS.R
Comment (jds485) [inline 2_process/src/raster_per_PRMS.R:73]: should this be an else or an else if? else seels like it would work

File: 2_process/src/raster_per_PRMS.R
Comment (jds485) [inline 2_process/src/raster_per_PRMS.R:15]: Could be helpful to add `stop` conditions for raster summary being NULL when categorical is NULL or FALSE. Would this function work if the prefix is NULL? Seems like `paste` treats NULL as a space, so it's okay if it's left as NULL.

File: 2_process.R
Comment (msleckman) [reply 2_process.R:70]: yes only for year 1960 . Mapping not included yet. I plan to adjust this target to loop through all the files in `p1_backcasted_LC`.

File: 2_process/src/raster_per_PRMS.R
Comment (msleckman) [reply 2_process/src/raster_per_PRMS.R:15]: Yea `paste()` treats `NULL` as nothing. So I planned to leave it in, the col name will simply have no prefix in `new_col_prefix` is `NULL`

File: 2_process/src/raster_per_PRMS.R
Comment (msleckman) [reply 2_process/src/raster_per_PRMS.R:36]: Yes trying to see what's happening here. Also becoming an issue for gridmet. seem new issue #54

**Comment by jds485** [comment]
I missed these comments yesterday:
> For non-discrete raster, the summary function is set to sum. Is this what we want per PRMS catchment for non-discrete rasters such as rd salt? 

Yes, sum is what we'd want for road salt and for atmospheric deposition. Meteorological data would use mean.

> Moreover, I am not sure how this aggregation is working with terra::extract, especially given the low res of the rd salt data set.

Is your concern that the extraction could be based on the raster grid cell center instead of the raster's equivalent polygon grid? That would be good to check.

> For LC raster, the calculation produces a proportion calculation - # pixels/ total pixels per catchment. Need more calculation steps to get final result of lc class area/total area.

The pixels are approximately the same size, so I'd expect the area would essentially cancel out in the numerator and denominator.

**Comment by msleckman** [comment]
@lekoenig you can go ahead and review this pr now 

*Latest updates:* 
1. edits from Jared's comments
2. Updated the `download_tifs()` function to fetch both backcasted LC data and road salt
3. added `lapply()` to process rasters. 
4. renamed functions and R files

*To do:*
1. Have not had a chance to check the the calculations. as such the road salt aggregation calculation is still just a sum, when it should be sum over all years per catchment / all years entire drb. Right now the output p2_rd_salt is just a list of sf dataframes (15 dfs !!!) which is the catchments df expanded by one col to include the specific year rd salt sum at catchment level. A additional computation should immediately follow `p2_rd_salt` target creating a new df with aggregated rd salt across all years.
2. This branch is currently behind USGS-R main with merge conflicts. Will resolve asap before merging after PR is approved. You can still review PR.

**Comment by lekoenig** [comment]
@msleckman Do you mind linking the issues that are addressed in this PR?

**Review by lekoenig — CHANGES_REQUESTED** [review]
Thanks, @msleckman! It's awesome that these functions are generalizable across a few of the raster datasets we're after 😃 

I've made some minor comments here, mostly requests for documentation. However, I can't seem to build the two targets within the process phase file. Can you see my comment and let me know what you think? Perhaps I should try updating or re-installing terra, but I'm curious if those are building for you. Thanks!

File: 1_fetch.R
Comment (msleckman) [reply 1_fetch.R:150]: Yes for the model, time span if only between 1979-current. And FORE-SCE is backcasted available for every decade between 1680 and 2010. Here we select the years that are relevant for our model, and we will already have NLCD data for 2000 to 2010. 
Looking back at our original discussion about FORE-SCE we decided that we should grab 1960 as well. Is that still OK? Or should we rm 1960 as well. 
@jds485

File: _targets.R
Comment (msleckman) [reply _targets.R:101]: Yea I am fine with this. I've modified it such that we no longer have sb id and sb file name as a var in the `_targets.R` and instead we have those strings directly in `1_fetch.R`

File: 1_fetch.R
Comment (jds485) [reply 1_fetch.R:150]: We'll use the 1960 data for the ~20 yr lagged land cover time series

File: 1_fetch/src/download_tifs_annual.R
Comment (msleckman) [reply 1_fetch/src/download_tifs_annual.R:37]: So `file.remove()` will remove all the years I don't want in the download folder. `rm()` deletes the variables I no longer need. This is technically unnecessary as these are just local variables.

File: 2_process.R
Comment (msleckman) [reply 2_process.R]: > When x is specified as `p1_backcasted_LC[1]` does that mean that we're only calculating the raster land cover using the first tif, e.g.
Yes, however in my latest code push I've removed `[1]`, it was just kept to speed up the processing time. When removed, the two processed raster targets are lists that look like the following: 

image

image

These will be merged in the next process pR I am working on. 

@lekoenig  this builds for me successfully for me. But this crs error might be due to the catchments polygon having errors. Are you able to pull this and re-build to see if it works. Maybe just test the first tif file as it was originally when you reviewed:
```
tar_target(
    p2_FORESCE_LC_per_catchment, 
    {lapply(p1_backcasted_LC[1], function(x) raster_to_catchment_polygons(polygon_sf = p1_catchments_sf_valid,
                                                                          raster = x, categorical_raster = TRUE,
                                                                          raster_summary_fun = NULL, new_cols_prefix = 'lcClass'))
      }
  ),
  
  # Extract Road Salt raster values to catchments polygons in the DRB - general function raster_to_catchment_polygons
  tar_target(
    p2_rdsalt_per_catchment,
    {lapply(p1_rd_salt[1], function(x) raster_to_catchment_polygons(polygon_sf = p1_catchments_sf_valid,
                                                                    raster = x, categorical_raster = FALSE,
                                                                    raster_summary_fun = sum, new_cols_prefix = 'rd_slt'))
    }
  ), 
```

File: 1_fetch.R
Comment (lekoenig) [reply 1_fetch.R:160]: Thanks for adding these comments that give some context to the `year` argument for both `p1_backcasted_LC` and `p1_rd_salt`!

File: 2_process/src/raster_to_catchment_polygons.R
Comment (lekoenig) [inline 2_process/src/raster_to_catchment_polygons.R:10]: Do you mind adding some documentation for this function, i.e. what is it doing?

File: 2_process/src/raster_to_catchment_polygons.R
Comment (lekoenig) [inline 2_process/src/raster_to_catchment_polygons.R:23]: Should `@param categorical_raster_prefix` be replaced with `new_cols_prefix` here, or is that a separate parameter?

File: 2_process/src/raster_to_catchment_polygons.R
Comment (lekoenig) [inline 2_process/src/raster_to_catchment_polygons.R:73]: Was this `print` statement mostly for exploratory purposes? Is it OK to remove it to clean up the pipeline output?

File: 2_process/src/raster_to_catchment_polygons.R
Comment (lekoenig) [inline 2_process/src/raster_to_catchment_polygons.R:81]: Do you think we'll need the (multi)polygons in a downstream step? If not, the output `df` might be cleaner if you ran `st_drop_geometry` on `vector_sf` either on line 79 or in within the join.

File: 1_fetch/src/download_tifs_annual.R
Comment (lekoenig) [reply 1_fetch/src/download_tifs_annual.R]: Ok, I see. Thanks.

**Comment by lekoenig** [comment]
Thanks, Margaux! The code now runs for me 👍 

> Have not had a chance to edit the calculations happening when aggregating to the catchment polygon. As such, the road salt aggregation calculation is still just a sum (raster_to_catchment_polygons.R line:75), when it should ultimately be a sum over all years per catchment / all years entire drb.

I haven't really dug in to the raster processing calculations at all here. Is the idea that you'll edit these calculations in a future PR? If that's the case, I've made a couple of minor comments but approve of this PR once you've evaluated those.

**Review by lekoenig — APPROVED** [review]

File: 2_process/src/raster_to_catchment_polygons.R
Comment (msleckman) [reply 2_process/src/raster_to_catchment_polygons.R:23]: yes this was old documentation - pre rd salt. thanks!

File: 2_process/src/raster_to_catchment_polygons.R
Comment (msleckman) [reply 2_process/src/raster_to_catchment_polygons.R:89]: ah yes i've removed the geometry col. Much cleaner. Thanks!

**Pull request merged**
